File: src/main/ReverseWordsInAString.java
package main;

/**
 Given an input string, reverse the string word by word.

 For example,
 Given s = "the sky is blue",
 return "blue is sky the".
 */
public class ReverseWordsInAString {

    public String reverseWords(String s) {
        s = s.trim(); //get rid of spaces
        if(s.length() == 0)
            return s;

        String result = "";
        int st = 0;
        int end = 0;
        while(end < s.length()) {

            if(s.charAt(st) != ' ') {
                if(s.charAt(end) == ' ' ) {
                    //only store if st != ' ' and end == ' '
                    result = s.substring(st,end) + " " + result;
                    st = end+1;
                }
            } else {
                //move st if is pointing to empty
                st++;
            }
            end++;
        }
        result = s.substring(st,end) + " " + result;

        return result.trim(); //in case of single word
    }
}

File: python/reverseWordsInAString.py
__author__ = 'don'
"""
Given an input string, reverse the string word by word.

For example,
Given s = "the sky is blue",
return "blue is sky the".

Update (2015-02-12):
For C programmers: Try to solve it in-place in O(1) space.

"""
class Solution:
    # @param s, a string
    # @return a string
    def reverseWords(self, s):
        return ' '.join(reversed(s.split()))





if __name__ == '__main__':
    print Solution().reverseWords('the sky is blue')
